### PR TITLE
*: support create table with range type table partition.

### PR DESCRIPTION
**Pull request by coocood**
The first step to implement table partition.

### CODE REVIEW

**Comment by shenli** [comment]
@coocood We should merge this PR after the 2.0-GA release.

File: model/model.go [hunk 1]
```diff
@@ -221,6 +223,34 @@ func (t *TableInfo) ColumnIsInIndex(c *ColumnInfo) bool {
...
+	Columns []CIStr
+
+	Definitions []PartitionDefinition
+}
+
+// PartitionDefinition defines a single partition.
+type PartitionDefinition struct {
+	ID       int64
+	Name     string
+	LessThan []string
```
Comment (zz-jason) [inline model/model.go]: `LessThan` is only used in range partition ?

File: ast/ddl.go [hunk 2]
```diff
@@ -846,3 +847,19 @@ func (n *TruncateTableStmt) Accept(v Visitor) (Node, bool) {
 	n.Table = node.(*TableName)
 	return v.Leave(n)
 }
+
+// PartitionDefinition defines a single partition.
+type PartitionDefinition struct {
```
[the remark this answers is not in the document]
Comment (coocood) [reply ast/ddl.go]: It can hold other partition types.

File: model/model.go (hunk 1 above)
Comment (coocood) [reply model/model.go]: Yes

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -780,6 +788,42 @@ func (d *ddl) CreateTable(ctx sessionctx.Context, ident ast.Ident, colDefs []*as
...
+				return errors.Trace(err1)
+			}
+			piDef := model.PartitionDefinition{
+				Name:     def.Name,
+				ID:       pid,
+				Comment:  def.Comment,
+				MaxValue: def.MaxValue,
+			}
+			for _, expr := range def.LessThan {
+				var buf = bytes.NewBuffer([]byte{})
```
Comment (tiancaiamao) [inline ddl/ddl_api.go]: ditto

**Comment by tiancaiamao** [comment]
LGTM

File: model/model.go (hunk 1 above)
Comment (zz-jason) [reply model/model.go]: so how do we define a hash or list or key partition ?

Comment (tiancaiamao) [reply model/model.go]: It would be implemented later... @zz-jason

Comment (zz-jason) [reply model/model.go]: I think we should implement a more extensible struct to define partition info.

Comment (coocood) [reply model/model.go]: It's compatible when we add new fields later.

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -740,14 +740,22 @@ func (d *ddl) CreateTableWithLike(ctx sessionctx.Context, ident, referIdent ast.
...
+		return d.CreateTableWithLike(ctx, ident, referIdent)
+	}
+	colDefs := s.Cols
 	is := d.GetInformationSchema()
 	schema, ok := is.SchemaByName(ident.Schema)
 	if !ok {
 		return infoschema.ErrDatabaseNotExists.GenByArgs(ident.Schema)
 	}
 	if is.TableExists(ident.Schema, ident.Name) {
+		if s.IfNotExists {
```
Comment (zimulala) [inline ddl/ddl_api.go]: Could you add a test to cover this logic?

**Pull request merged**